ompi/include/mpi.h.in

```diff
@@ -1742,14 +1742,18 @@ OMPI_DECLSPEC  int MPI_Info_delete(MPI_Info info, const char *key);
 OMPI_DECLSPEC  int MPI_Info_dup(MPI_Info info, MPI_Info *newinfo);
 OMPI_DECLSPEC  MPI_Info MPI_Info_f2c(MPI_Fint info);
 OMPI_DECLSPEC  int MPI_Info_free(MPI_Info *info);
+#if MPI_VERSION < 4
```

Comment:
It might be good to put a comment here (and all the `#if MPI_VERSION < 4` blocks) that even for MPI_VERSION<4, the function *is* prototyped; it's just prototyped below with the deprecated warning.

Reply:
How about:
```suggestion
/*
 * In MPI 4.0 this function is deprecated but not (yet) removed.
 * As such, this function is defined with the deprecation warning later
 * in this file. It is grouped with some of the other deprecated functions.
 */
#if MPI_VERSION < 4
```
If that works, they @hppritcha can apply the comment to the other `#if MPI_VERSION < 4` blocks. Then I think we are ready to merge.

Reply:
Thinking about this a bit more, is there a way we can avoid duplicating the prototypes of the deprecated functions?  It's not a deal-breaker if we can't, but I just hate replicating code...

Reply:
The best I could think of would look something like:
```c
/*
 * In MPI 4.0 this function is deprecated but not (yet) removed.
 * As such, this function is defined with the deprecation warning for
 * builds of Open MPI with compliance levels of 4.0 and later.
 */
OMPI_DECLSPEC  int MPI_Info_get(MPI_Info info, const char *key, int valuelen,
                                char *value, int *flag)
#if MPI_VERSION < 4
            __mpi_interface_deprecated__("MPI_Info_get was deprecated in MPI-4.0; use MPI_Info_get_string instead")
#endif
;
```
I don't love it though (mostly because of the trailing `;`).

Reply:
How about:

```c
#if MPI_VERSION < 4
#define MPI4_DEPRECATED(msg) \
    __attribute__ ((deprecated (msg) ));
#else
#define MPI4_DEPRECATED(name, instead)
#endif

OMPI_DECLSPEC  int MPI_Info_get(MPI_Info info, const char *key, int valuelen,
                                char *value, int *flag)
    MPI4_DEPRECATED("MPI_Info_get was deprecated in MPI-4.0; use MPI_Info_get_string instead");
```

Reply:
I can tinker around and try something like that. I'll post a PR in a few.

If we go back to 1 interface with this technique, then we should probably move these into the deprecated section, right? We seem to be collecting them towards the bottom of the file.

Reply:
> If we go back to 1 interface with this technique, then we should probably move these into the deprecated section, right? We seem to be collecting them towards the bottom of the file.

Works for me; my goal is just to list them once -- I don't have strong opinions on where that "once" should be.  But I do like the suggestion of grouping all the MPI-4 deprecated prototypes at the bottom.

Reply:
@jsquyres What do you think about this:
 * https://github.com/hppritcha/ompi/pull/35

Reply:
@hppritcha if you agree, feel free to merge in my PR to your branch. You can also squash these all down to 1 (just add the co-author lines for Dave and I). Thanks!